Reduce use of WTF_ALLOW_UNSAFE_BUFFER in WebCore/rendering

cdumez opened the pull request:
#### 911d0d32b1e8f02d8dd8dd34688359018b5d3ec6
<pre>
Reduce use of WTF_ALLOW_UNSAFE_BUFFER in WebCore/rendering
<a href="https://bugs.webkit.org/show_bug.cgi?id=283780">https://bugs.webkit.org/show_bug.cgi?id=283780</a>

Reviewed by Geoffrey Garen.

* Source/WebCore/rendering/RenderTableSection.cpp:
* Source/WebCore/rendering/StyledMarkedText.cpp:
* Source/WebCore/rendering/TextDecorationPainter.cpp:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::cssWeightOfSystemFont):
* Source/WebCore/rendering/line/BreakingContext.h:
* Source/WebCore/rendering/mac/RenderThemeMac.h:
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
(WebCore::inflateRect):
(WebCore::controlSizeFromPixelSize):
(WebCore::popupButtonMargins):
(WebCore::popupButtonSizes):
(WebCore::popupButtonPadding):
(WebCore::sizeForFont):
(WebCore::sizeForSystemFont):
(WebCore::setSizeFromFont):
(WebCore::menuListButtonSizes):
(WebCore::RenderThemeMac::popupInternalPaddingBox const):
(WebCore::RenderThemeMac::menuListSizes const):
(WebCore::RenderThemeMac::searchFieldSizes const):
(WebCore::RenderThemeMac::cancelButtonSizes const):
(WebCore::RenderThemeMac::resultsButtonSizes const):
* Source/WebCore/rendering/mathml/MathOperator.cpp:
* Source/WebCore/rendering/style/QuotesData.cpp:
* Source/WebCore/rendering/style/TextSizeAdjustment.cpp:

Canonical link: <a href="https://commits.webkit.org/287157@main">https://commits.webkit.org/287157@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/96bba88023438e875bbd98723746cf5e48d3c560

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78549 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57594 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31931 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83210 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29814 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80682 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66745 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5875 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61524 "Found 1 new test failure: media/modern-media-controls/css/transformed-media-crash.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19441 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81616 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51535 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/69787 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41835 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48881 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/25367 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28151 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69981 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/25740 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84576 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5914 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4051 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69749 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6075 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67511 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69003 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13025 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/11468 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12129 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5861 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/11757 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5849 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9283 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7636 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -180,36 +178,37 @@ static Vector<StyledMarkedText> coalesceAdjacentWithSameRanges(Vector<StyledMark
     frontmostMarkedTexts.append(styledTexts[0]);
-    for (auto it = styledTexts.begin() + 1, end = styledTexts.end(); it != end; ++it) {
-        StyledMarkedText& previousStyledMarkedText = frontmostMarkedTexts.last();
+    for (size_t i = 1; i < styledTexts.size(); ++i) {
```
This can be a ranged for loop.

cdumez replied:
How? I would need a boolean variable to skip the first item I think? If so, it didn't feel better to me. Is there a better way to write this?

geoffreygaren replied:
Oops! I didn't notice the 1. Yeah, I think you're right.

geoffreygaren commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -288,14 +287,15 @@ Vector<StyledMarkedText> StyledMarkedText::subdivideAndResolve(const Vector<Mark
     frontmostMarkedTexts.append(resolveStyleForMarkedText(markedTexts[0], baseStyle, renderer, lineStyle, paintInfo));
-    for (auto it = markedTexts.begin() + 1, end = markedTexts.end(); it != end; ++it) {
-        StyledMarkedText& previousStyledMarkedText = frontmostMarkedTexts.last();
+    for (size_t i = 1; i < markedTexts.size(); ++i) {
```
Ditto

geoffreygaren commented on Source/WebCore/rendering/StyledMarkedText.cpp:
```diff
@@ -314,13 +314,14 @@ static Vector<StyledMarkedText> coalesceAdjacent(const Vector<StyledMarkedText>&
-        StyledMarkedText& previousStyledMarkedText = styledMarkedTexts.last();
-        if (areAdjacentMarkedTextsWithSameStyle(previousStyledMarkedText, *it)) {
-            previousStyledMarkedText.endOffset = it->endOffset;
+    for (size_t i = 1; i < textsToCoalesce.size(); ++i) {
```
Ditto

geoffreygaren commented on Source/WebCore/rendering/TextDecorationPainter.cpp:
```diff
@@ -123,24 +121,23 @@ static DashArray translateIntersectionPointsToSkipInkBoundaries(const DashArray&
-        intermediateTuples.append(*tuples.begin());
-        for (auto i = tuples.begin() + 1; i != tuples.end(); i++) {
+        intermediateTuples.append(tuples.first());
+        for (size_t i = 1; i < tuples.size(); ++i) {
```
Ditto

geoffreygaren commented on Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
```diff
@@ -241,8 +239,8 @@ static inline FontSelectionValue cssWeightOfSystemFont(CTFontRef font)
-    static constexpr float weightThresholds[] = { -0.6, -0.365, -0.115, 0.130, 0.235, 0.350, 0.5, 0.7 };
-    for (unsigned i = 0; i < std::size(weightThresholds); ++i) {
+    static constexpr std::array weightThresholds { -0.6f, -0.365f, -0.115f, 0.130f, 0.235f, 0.350f, 0.5f, 0.7f };
+    for (unsigned i = 0; i < weightThresholds.size(); ++i) {
```
Dito

geoffreygaren commented on Source/WebCore/rendering/mac/RenderThemeMac.mm:
```diff
@@ -750,39 +748,36 @@ static NSControlSize controlSizeFromPixelSize(const IntSize* sizes, const IntSiz
-        { 0, 3, 2, 3 },
-        { 0, 1, 0, 1 },
-        { 0, 6, 2, 6 },
+    static constexpr std::array<std::array<int, 4>, 4> margins {
```
I think it’s a little nicer to use auto and std::to_array, so you don’t have to write out the ugly nested std::array declaration, and if someone adds or removes an entry they don’t have to update the type declaration.

geoffreygaren commented on Source/WebCore/rendering/mac/RenderThemeMac.mm:
```diff
@@ -750,39 +748,36 @@ static NSControlSize controlSizeFromPixelSize(const IntSize* sizes, const IntSiz
-        { 2, 23, 3, 8 },
-        { 2, 22, 3, 10 },
-        { 2, 26, 3, 8 },
+    static constexpr std::array<std::array<int, 4>, 4> paddingLTR {
```
Ditto

geoffreygaren commented on Source/WebCore/rendering/mac/RenderThemeMac.mm:
```diff
@@ -750,39 +748,36 @@ static NSControlSize controlSizeFromPixelSize(const IntSize* sizes, const IntSiz
+static std::span<const IntSize, 4> popupButtonSizes()
 {
-    static const IntSize sizes[4] = { IntSize(0, 21), IntSize(0, 18), IntSize(0, 15), IntSize(0, 24) };
+    static constexpr std::array<IntSize, 4> sizes { IntSize(0, 21), IntSize(0, 18), IntSize(0, 15), IntSize(0, 24) };
```
Ditto

geoffreygaren commented on Source/WebCore/rendering/mac/RenderThemeMac.mm:
```diff
@@ -750,39 +748,36 @@ static NSControlSize controlSizeFromPixelSize(const IntSize* sizes, const IntSiz
-        { 2, 8, 3, 23 },
-        { 2, 8, 3, 22 },
-        { 2, 8, 3, 26 },
+    static constexpr std::array<std::array<int, 4>, 4> paddingRTL {
```
Ditto

geoffreygaren commented on Source/WebCore/rendering/mac/RenderThemeMac.mm:
```diff
@@ -963,9 +958,9 @@ static void setFontFromControlSize(RenderStyle& style, NSControlSize controlSize
+static std::span<const IntSize, 4> menuListButtonSizes()
 {
-    static const IntSize sizes[4] = { IntSize(0, 21), IntSize(0, 18), IntSize(0, 15), IntSize(0, 28) };
+    static constexpr std::array<IntSize, 4> sizes { IntSize(0, 21), IntSize(0, 18), IntSize(0, 15), IntSize(0, 28) };
```
Ditto

geoffreygaren commented on Source/WebCore/rendering/mac/RenderThemeMac.mm:
```diff
@@ -1050,9 +1045,9 @@ static void setFontFromControlSize(RenderStyle& style, NSControlSize controlSize
+std::span<const IntSize, 4> RenderThemeMac::menuListSizes() const
 {
-    static const IntSize sizes[4] = { IntSize(9, 0), IntSize(5, 0), IntSize(0, 0), IntSize(13, 0) };
+    static constexpr std::array<IntSize, 4> sizes { IntSize(9, 0), IntSize(5, 0), IntSize(0, 0), IntSize(13, 0) };
```
Ditto

geoffreygaren commented on Source/WebCore/rendering/mac/RenderThemeMac.mm:
```diff
@@ -1072,9 +1067,9 @@ static void setFontFromControlSize(RenderStyle& style, NSControlSize controlSize
+std::span<const IntSize, 4> RenderThemeMac::searchFieldSizes() const
 {
-    static const IntSize sizes[4] = { IntSize(0, 22), IntSize(0, 19), IntSize(0, 17), IntSize(0, 30) };
+    static constexpr std::array<IntSize, 4> sizes { IntSize(0, 22), IntSize(0, 19), IntSize(0, 17), IntSize(0, 30) };
```
Ditto

geoffreygaren commented on Source/WebCore/rendering/mac/RenderThemeMac.mm:
```diff
@@ -1120,9 +1115,9 @@ static void setFontFromControlSize(RenderStyle& style, NSControlSize controlSize
+std::span<const IntSize, 4> RenderThemeMac::cancelButtonSizes() const
 {
-    static const IntSize sizes[4] = { IntSize(22, 22), IntSize(19, 19), IntSize(15, 15), IntSize(22, 22) };
+    static constexpr std::array<IntSize, 4> sizes { IntSize(22, 22), IntSize(19, 19), IntSize(15, 15), IntSize(22, 22) };
```
Ditto

geoffreygaren commented on Source/WebCore/rendering/mac/RenderThemeMac.mm:
```diff
@@ -1134,10 +1129,10 @@ static void setFontFromControlSize(RenderStyle& style, NSControlSize controlSize
+std::span<const IntSize, 4> RenderThemeMac::resultsButtonSizes() const
 {
-    static const IntSize sizes[4] = { IntSize(19, 22), IntSize(17, 19), IntSize(17, 15), IntSize(19, 22) };
+    static constexpr std::array<IntSize, 4> sizes { IntSize(19, 22), IntSize(17, 19), IntSize(17, 15), IntSize(19, 22) };
```
Ditto

geoffreygaren reviewed (approved):
r=me

cdumez committed:
Reduce use of WTF_ALLOW_UNSAFE_BUFFER in WebCore/rendering
https://bugs.webkit.org/show_bug.cgi?id=283780

Reviewed by Geoffrey Garen.

* Source/WebCore/rendering/RenderTableSection.cpp:
* Source/WebCore/rendering/StyledMarkedText.cpp:
* Source/WebCore/rendering/TextDecorationPainter.cpp:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::cssWeightOfSystemFont):
* Source/WebCore/rendering/line/BreakingContext.h:
* Source/WebCore/rendering/mac/RenderThemeMac.h:
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
(WebCore::inflateRect):
(WebCore::controlSizeFromPixelSize):
(WebCore::popupButtonMargins):
(WebCore::popupButtonSizes):
(WebCore::popupButtonPadding):
(WebCore::sizeForFont):
(WebCore::sizeForSystemFont):
(WebCore::setSizeFromFont):
(WebCore::menuListButtonSizes):
(WebCore::RenderThemeMac::popupInternalPaddingBox const):
(WebCore::RenderThemeMac::menuListSizes const):
(WebCore::RenderThemeMac::searchFieldSizes const):
(WebCore::RenderThemeMac::cancelButtonSizes const):
(WebCore::RenderThemeMac::resultsButtonSizes const):
* Source/WebCore/rendering/mathml/MathOperator.cpp:
* Source/WebCore/rendering/style/QuotesData.cpp:
* Source/WebCore/rendering/style/TextSizeAdjustment.cpp:

Canonical link: https://commits.webkit.org/287157@main
